### PR TITLE
Move to chrono

### PR DESCRIPTION
**Pull request by alexbriskin**
## Description
Use standard C++ time measurements and waits instead of rdtsc.

##### What
_Subject: what this PR is doing in one line._ 

##### Why ?
_Justification for the PR. If there is existing issue/bug please reference._

##### How ?
_It is optional but for complex PRs please provide information about the design,
architecture, approach, etc._

## Change type
What kind of change does this PR introduce?
- [ ] Bugfix
- [ ] Feature
- [ ] Code style update
- [X] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Tests
- [ ] Other

## Check list
- [ ] Code follows the style de facto guidelines of this project
- [ ] Comments have been inserted in hard to understand places
- [ ] Documentation has been updated (if necessary)
- [ ] Test has been added (if possible)

### CODE REVIEW

**Comment by alexbriskin** [comment]
bot:retest

**Comment by alexbriskin** [comment]
bot:retest

**Comment by alexbriskin** [comment]
bot:retest

**Comment by alexbriskin** [comment]
bot:retest

**Comment by alexbriskin** [comment]
bot:retest

**Comment by galnoam** [comment]
@iftahl please review the PR, thanks.

File: src/utils/rdtsc.h [hunk 1]
```diff
@@ -117,42 +119,13 @@ static inline tscval_t get_tsc_rate_per_second()
...
-
-    return 0;
-}
-
 static inline int gettime(struct timespec *ts)
 {
-    return gettimefromtsc(ts);
+    auto now = steady_clock::now().time_since_epoch();
+    auto now_seconds = duration_cast<seconds>(now);
+    ts->tv_sec = now_seconds.count();
```
Comment (iftahl) [inline src/utils/rdtsc.h:75]: what happens when tv_sec is only 32 bit field?
happens when "__USE_TIME_BITS64" is not defined

File: src/vlogger/vlogger.cpp [hunk 2]
```diff
@@ -32,6 +32,7 @@
 
 #include "vlogger.h"
 
+#include <chrono>
```
Comment (iftahl) [inline src/vlogger/vlogger.cpp:35]: probably not needed since its in the header, right?

File: src/vlogger/vlogger.cpp [hunk 3]
```diff
@@ -48,14 +49,16 @@
 #define VLOG_DEFAULT_MODULE_NAME "XLIO"
 #define XLIO_LOG_CB_ENV_VAR      "XLIO_LOG_CB_FUNC_PTR"
 
+using namespace std::chrono;
```
Comment (iftahl) [inline src/vlogger/vlogger.cpp:52]: probably not needed since its in the header, right?

File: src/vlogger/vlogger.h [hunk 4]
```diff
@@ -359,20 +361,13 @@ void vlog_stop(void);
...
+    auto now = steady_clock::now();
 
-    if (!g_vlogger_usec_on_startup) {
-        g_vlogger_usec_on_startup = ts_to_usec(&ts_now);
+    if (g_vlogger_usec_on_startup.time_since_epoch() == steady_clock::duration::zero()) {
+        g_vlogger_usec_on_startup = now;
     }
 
-    return (ts_to_usec(&ts_now) - g_vlogger_usec_on_startup);
+    return duration_cast<microseconds>(now - g_vlogger_usec_on_startup).count();
```
Comment (iftahl) [inline src/vlogger/vlogger.h:370]: does uint32_t is enough to hold us.count()?
what heppens if .count() is > max uint32_t?

its a general question for all the code.

File: src/core/main.cpp [hunk 5]
```diff
@@ -298,6 +298,57 @@ void check_debug()
...
+        double mhz = 0;
+        int rc = 0;
+
+#if defined(__ia64__)
+        rc = sscanf(buf, "itc MHz : %lf", &mhz);
+#elif defined(__powerpc__)
+        rc = sscanf(buf, "clock : %lf", &mhz);
+#elif defined(__aarch64__)
+        rc = sscanf(buf, "BogoMIPS : %lf", &mhz);
+        mhz /= 2;
```
Comment (iftahl) [inline src/core/main.cpp:327]: why mhz is half of BogoMIPS?

**Review by iftahl — COMMENTED** [review]
see comments & questions

File: src/core/main.cpp (hunk 5 above)
Comment (iftahl) [reply src/core/main.cpp:327]: on one of our bf3 for example:
BogoMIPS	: 660.55

but:
	Max Speed: 2134 MHz
	Current Speed: 2134 MHz
as you can see from "dmidecode -t 4" output

**Comment by galnoam** [comment]
@alexbriskin , please fix/comment on iftah review.
Thanks.

File: src/stats/stats_reader.cpp [hunk 6]
```diff
@@ -1585,15 +1583,12 @@ void stats_reader_handler(sh_mem_t *p_sh_mem, int pid)
...
         }
-        if (gettime(&end)) {
-            log_system_err("gettime()");
-            goto out;
-        }
-        uint64_t delay_int_micro = SEC_TO_MICRO(user_params.interval);
-        uint64_t adjasted_delay = delay_int_micro - TIME_DIFF_in_MICRO(start, end);
+        auto end = steady_clock::now();
+        auto duration = std::chrono::seconds(user_params.interval);
+        auto adjasted_duration = duration - (start - end);
```
[the remark this answers is not in the document]
Comment (alexbriskin) [reply src/stats/stats_reader.cpp:1588]:
```suggestion
        auto adjasted_duration = duration - (end - start);
```

File: src/utils/rdtsc.h (hunk 1 above)
Comment (alexbriskin) [reply src/utils/rdtsc.h:75]: In old 32-bit systems where the time_t does not contain at least 35-bits, two things might happen:

- Compiler will warn about the value truncation.
- In runtime, the two instances(start, end) of gettime will still be comparable. Unless a wraparound occurred in runtime(extremely unlikely) there will be no issue.

File: src/vlogger/vlogger.cpp (hunk 3 above)
Comment (alexbriskin) [reply src/vlogger/vlogger.cpp:52]: We can remove this line but then the following line:
`time_point<steady_clock> g_vlogger_usec_on_startup;`
will look as follows:
`std::chrono::time_point<std::chrono::steady_clock> g_vlogger_usec_on_startup;`
Is this better?

File: src/vlogger/vlogger.h (hunk 4 above)
Comment (alexbriskin) [reply src/vlogger/vlogger.h:370]: > does uint32_t is enough to hold us.count()?

No

> what heppens if .count() is > max uint32_t?

At worst the comiler will look at the 32 least significant bits, same as now.

File: src/vlogger/vlogger.cpp (hunk 3 above)
Comment (iftahl) [reply src/vlogger/vlogger.cpp:52]: no, better to use namespace. just wondered if we need it here at all... nevermind

File: src/core/main.cpp (hunk 5 above)
Comment (iftahl) [reply src/core/main.cpp:327]: we must solve it here...
in case if defined(__aarch64__) is true - "mhz" will be incorrect

Comment (iftahl) [reply src/core/main.cpp:327]: check if we can count on /sys/devices/system/cpu/cpu0/cpufreq/cpuinfo_max_freq
or maybe
/sys/devices/system/cpu/cpu0/cpufreq/cpuinfo_cur_freq

File: src/core/util/agent.cpp [hunk 7]
```diff
@@ -378,40 +381,33 @@ int agent::put(const void *data, size_t length, intptr_t tag)
...
-        /* cppcheck-suppress syntaxError */
-        if (tv_cmp(&tv_inactive_elapsed, &tv_now, <)) {
-            tv_inactive_elapsed = tv_now;
-            tv_inactive_elapsed.tv_sec += AGENT_DEFAULT_INACTIVE;
-            if (0 <= send_msg_init()) {
-                progress_cb();
-                goto go;
-            }
+    /* Attempt can be done less often than progress in active state */
+    if (AGENT_INACTIVE == m_state && inactive_elapsed < now) {
```
[the remark this answers is not in the document]
Comment (alexbriskin) [reply src/core/util/agent.cpp]: Got it. Fixed it!

File: src/core/main.cpp (hunk 5 above)
Comment (alexbriskin) [reply src/core/main.cpp]: I won't make this code change in this PR.

**Review by iftahl — APPROVED** [review]

**Pull request closed without merging**